support/metrics/metricsTracker.go

```diff
@@ -0,0 +1,179 @@
+	commonProps
+	Exit       bool   `json:"exit"`
+	StackTrace string `json:"stack_trace"`
+}
+
+// MakeMetricsTracker is a factory method to create a `metrics.Tracker`.
+func MakeMetricsTracker(
+	userID string,
+	apiKey string,
+	client *http.Client,
+	start time.Time,
+	version string,
+	goos string,
+	goarch string,
+	goarm string,
+	guiVersion string,
+	strategy string,
+	updateTimeIntervalSeconds int32,
+	exchange string,
+	tradingPair string,
+) (*MetricsTracker, error) {
+	sessionID, e := uuid.NewUUID()
+	if e != nil {
+		return nil, fmt.Errorf("could not generate uuid with error %s", e)
+	}
+	props := commonProps{
+		CliVersion:                version,
+		Goos:                      goos,
+		Goarch:                    goarch,
+		Goarm:                     goarm,
+		GuiVersion:                guiVersion,
+		Strategy:                  strategy,
+		UpdateTimeIntervalSeconds: updateTimeIntervalSeconds,
+		Exchange:                  exchange,
+		TradingPair:               tradingPair,
+		SessionID:                 sessionID,
+	}
+
+	return &MetricsTracker{
+		userID: userID,
+		client: client,
+		apiKey: apiKey,
+		props:  props,
+		start:  start,
+	}, nil
+}
+
+// SendStartupEvent sends the startup Amplitude event.
+func (mt *MetricsTracker) SendStartupEvent() error {
+	return mt.sendEvent(startupEventName, mt.props)
+}
+
+// SendUpdateEvent sends the update Amplitude event.
+func (mt *MetricsTracker) SendUpdateEvent(now time.Time, success bool) error {
+	commonProps := mt.props
+	commonProps.SecondsSinceStart = now.Sub(mt.start).Seconds()
+	updateProps := updateProps{
+		commonProps: commonProps,
+		Success:     success,
+	}
+	return mt.sendEvent(updateEventName, updateProps)
+}
+
+// SendDeleteEvent sends the delete Amplitude event.
+func (mt *MetricsTracker) SendDeleteEvent(exit bool) error {
+	commonProps := mt.props
+	commonProps.SecondsSinceStart = time.Now().Sub(mt.start).Seconds()
+	deleteProps := deleteProps{
+		commonProps: commonProps,
+		Exit:        exit,
+		StackTrace:  string(debug.Stack()),
+	}
+
+	return mt.sendEvent(deleteEventName, deleteProps)
+}
+
+// TODO DS Re-implement using `networking/JSONRequestDynamicHeaders`.
+func (mt *MetricsTracker) sendEvent(eventType string, eventProps interface{}) error {
+	requestBody, e := json.Marshal(map[string]interface{}{
+		"api_key": mt.apiKey,
+		"events": []event{event{
+			UserID:    mt.userID,
+			DeviceID:  mt.userID,
+			EventType: eventType,
+			Props:     eventProps,
+		}},
+	})
+
+	if e != nil {
+		log.Print("could not marshal json request")
+		return fmt.Errorf("could not marshal json request: %s", e)
+	}
+
+	log.Printf("Sending request: %v\n", string(requestBody))
+
+	var responseData interface{}
+	e = networking.JSONRequest(mt.client, "POST", amplitudeAPIURL, string(requestBody), map[string]string{}, &responseData, "")
+	if e != nil {
+		return fmt.Errorf("could not post amplitude request: %s", e)
+	}
+
+	log.Printf("Got response: %v\n", responseData)
```

Comment:
@debnil There are a total of three log lines here:
1. sending request (line 169)
2. got response (line 175)
3. successfully send event (line 177)

Instead of logging all three, we should remove (1) and (2) and change (3) to do this:
```
if strings.Contains(string(responseData), "success") {
    log.Printf("Successfully sent event metric of type '%s'", eventType)
} else {
    log.Printf("Failed to send event metric of type '%s' (request=%s; response: %s)", eventType, string(requestBody), string(responseData))
}
```

This will reduce the overhead of logging in relation to these requests and will at the same time give us the relevant information when needed.

PS: in the current state we will log line 177 (success) even if the response data indicates a failure. this will be fixed by the above suggestion

Reply:
Great suggestion for space-saving, but the suggested implementation (and variants of it) don't work cleanly. `responseData` is of `type interface{}`, so converting it to a string needs a type assertion. I've slightly modified this to define `responseDataStr := fmt.Sprintf("%v", responseData)